### DISCUSSION TITLE
Unable to upload images to Google Cloud Storage (GCS)

### DISCUSSION THREAD

**Discussion by dheeban-k** [outline/outline#3426] [Self Hosting]
Can someone help us with understanding how the file/image uploads works in GCS. The docker image of outline is hosted in Cloud Run, where the GCS bucket details are passed as ENV's.

When we try to insert image/file uploads it throws error "Sorry, an error occurred uploading the image"

**Comment by tommoor** [comment]
I'm sure folks will be willing to help but you need to provide the logs from the server and client + your environment variables to enable effective debugging

**Reply by dheeban-k** [reply]
Sure, @tommoor 
This is the error message seen in client.
<img width="1791" alt="image" src="https://user-images.githubusercontent.com/101546615/164171172-b1572cc4-b462-44c4-aad1-c6a091b8ee39.png">

Whereas, in server no logs of error pops up. Same scenario while we try out in cloud run as well. 
<img width="1280" alt="Screenshot_2022-04-20_at_12 40 55_PM" src="https://user-images.githubusercontent.com/101546615/164171528-07d67b95-6985-4be4-bf8e-2fd7894146db.png">

**Reply by almereyda** [reply]
When you look at the network tab of your browser's developer tools, do you see a failing request that correlates with the error thrown at the user in the UI?

Could you also share with us a redacted copy of the environment variables of the application, with special regards to the storage configuration?

**Comment by dheeban-k** [comment]
@almereyda we dont't see a failure code in the networking window.
<img width="1510" alt="image" src="https://user-images.githubusercontent.com/101546615/164461244-21e8bf47-6755-4af7-9544-520568cc3dee.png">


Will attach the redacted version of the env vars shortly. Thanks for the response :)

**Reply by dheeban-k** [reply]
@almereyda the error logs seen in the console are:

uploadFile.ts:41 Refused to connect to 'localhost://getoutline-001' because it violates the following Content Security Policy directive: "connect-src *". Note that '*' matches only URLs with network schemes ('http', 'https', 'ws', 'wss'), or URLs whose scheme matches `self`'s scheme. localhost:' must be added explicitely.

_callee$ @ uploadFile.ts:41
tryCatch @ runtime.js:63
invoke @ runtime.js:293
(anonymous) @ runtime.js:118
asyncGeneratorStep @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
_next @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
Promise.then (async)
asyncGeneratorStep @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
_next @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
uploadFile @ uploadFile.ts:10
_callee$ @ Editor.tsx:40
tryCatch @ runtime.js:63
invoke @ runtime.js:293
(anonymous) @ runtime.js:118
asyncGeneratorStep @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
_next @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ Editor.tsx:38
_loop @ insertFiles.ts:82
insertFiles @ insertFiles.ts:62
(anonymous) @ CommandMenu.tsx:291
Rb @ react-dom.production.min.js:52
Xb @ react-dom.production.min.js:52
Yb @ react-dom.production.min.js:53
Ze @ react-dom.production.min.js:100
se @ react-dom.production.min.js:101
(anonymous) @ react-dom.production.min.js:113
Pt @ react-dom.production.min.js:292
Nb @ react-dom.production.min.js:50
jd @ react-dom.production.min.js:105
yc @ react-dom.production.min.js:75
hd @ react-dom.production.min.js:74
(anonymous) @ scheduler.production.min.js:18
gg @ react-dom.production.min.js:122
Hb @ react-dom.production.min.js:292
gd @ react-dom.production.min.js:73
Show 8 more frames
uploadFile.ts:41 Refused to connect to 'localhost://getoutline-001' because it violates the document's Content Security Policy.
_callee$ @ uploadFile.ts:41
tryCatch @ runtime.js:63
invoke @ runtime.js:293
(anonymous) @ runtime.js:118
asyncGeneratorStep @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
_next @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
Promise.then (async)
asyncGeneratorStep @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
_next @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
uploadFile @ uploadFile.ts:10
_callee$ @ Editor.tsx:40
tryCatch @ runtime.js:63
invoke @ runtime.js:293
(anonymous) @ runtime.js:118
asyncGeneratorStep @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
_next @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1
(anonymous) @ Editor.tsx:38
_loop @ insertFiles.ts:82
insertFiles @ insertFiles.ts:62
(anonymous) @ CommandMenu.tsx:291
Rb @ react-dom.production.min.js:52
Xb @ react-dom.production.min.js:52
Yb @ react-dom.production.min.js:53
Ze @ react-dom.production.min.js:100
se @ react-dom.production.min.js:101
(anonymous) @ react-dom.production.min.js:113
Pt @ react-dom.production.min.js:292
Nb @ react-dom.production.min.js:50
jd @ react-dom.production.min.js:105
yc @ react-dom.production.min.js:75
hd @ react-dom.production.min.js:74
(anonymous) @ scheduler.production.min.js:18
gg @ react-dom.production.min.js:122
Hb @ react-dom.production.min.js:292
gd @ react-dom.production.min.js:73
Show 8 more frames
insertFiles.ts:123 TypeError: Failed to fetch
    at _callee$ (uploadFile.ts:41:32)
    at tryCatch (runtime.js:63:40)
    at Generator.invoke [as _invoke] (runtime.js:293:22)
    at Generator.next (runtime.js:118:21)
    at asyncGeneratorStep (default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1:100108)
    at _next (default~authenticated-routes~shared-document.ad7ae7003cf7d674a17d.js:1:100354)
(anonymous) @ insertFiles.ts:123
Promise.catch (async)
_loop @ insertFiles.ts:122
insertFiles @ insertFiles.ts:62
(anonymous) @ CommandMenu.tsx:291
Rb @ react-dom.production.min.js:52
Xb @ react-dom.production.min.js:52
Yb @ react-dom.production.min.js:53
Ze @ react-dom.production.min.js:100
se @ react-dom.production.min.js:101
(anonymous) @ react-dom.production.min.js:113
Pt @ react-dom.production.min.js:292
Nb @ react-dom.production.min.js:50
jd @ react-dom.production.min.js:105
yc @ react-dom.production.min.js:75
hd @ react-dom.production.min.js:74
(anonymous) @ scheduler.production.min.js:18
gg @ react-dom.production.min.js:122
Hb @ react-dom.production.min.js:292
gd @ react-dom.production.min.js:73

**Reply by almereyda** [reply]
I think the other error looks more promising:

> TypeError: Failed to fetch, insertFiles.ts:123

Unfortunately I don't know enough about Outline's internals, to be able to provide more hints. I *believe* the 2nd, 3rd and 5th request have something to do with this.

It will be very interesting to review your redacted configuration at some point.

**Reply by dheeban-k** [reply]
```
@almereyda , attaching the redacted env values

====================================

# –––––––––––––––– REQUIRED ––––––––––––––––


# Generate a hex-encoded 32-byte random key. You should use `openssl rand -hex 32`

# in your terminal to generate a random value.

SECRET_KEY=value generated using openssl rand -hex 32


# Generate a unique random key. The format is not important but you could still use

# `openssl rand -hex 32` in your terminal to produce this.

UTILS_SECRET=value generated using openssl rand -hex 32 


# For production point these at your databases, in development the default

# should work out of the box.

DATABASE_URL=postgres://username:password@Private IP:port/DB_name

DATABASE_URL_TEST=

DATABASE_CONNECTION_POOL_MIN=

DATABASE_CONNECTION_POOL_MAX=

# Uncomment this to disable SSL for connecting to Postgres

# PGSSLMODE=disable

REDIS_URL=redis://privateIp:port


# URL should point to the fully qualified, publicly accessible URL. If using a

# proxy the port in URL and PORT may be different.

URL=https://domanin.com/

PORT=port


# See [documentation](docs/SERVICES.md) on running a separate collaboration

# server, for normal operation this does not need to be set.

COLLABORATION_URL=


# To support uploading of images for avatars and document attachments an

# s3-compatible storage must be provided. AWS S3 is recommended for redundency

# however if you want to keep all file storage local an alternative such as

# minio (https://github.com/minio/minio) can be used.


# A more detailed guide on setting up S3 is available here:

# => https://wiki.generaloutline.com/share/125de1cc-9ff6-424b-8415-0d58c809a40f

#

AWS_ACCESS_KEY_ID=HMAC access id generated from GCS 

AWS_SECRET_ACCESS_KEY=HMAC secret key generated from GCS 

AWS_REGION=us

AWS_S3_UPLOAD_BUCKET_URL=https://storage.googleapis.com/

AWS_S3_UPLOAD_BUCKET_NAME=bucket name

AWS_S3_UPLOAD_MAX_SIZE=26214400

AWS_S3_FORCE_PATH_STYLE=true

AWS_S3_ACL=private



# –––––––––––––– AUTHENTICATION –––––––––––––


# To configure Google auth, you'll need to create an OAuth Client ID at

# => https://console.cloud.google.com/apis/credentials

#

# When configuring the Client ID, add an Authorized redirect URI:

GOOGLE_CLIENT_ID="client id from GCP"

GOOGLE_CLIENT_SECRET="client secret from GCP"


# –––––––––––––––– OPTIONAL ––––––––––––––––


# Base64 encoded private key and certificate for HTTPS termination. This is only

# required if you do not use an external reverse proxy. See documentation:

# https://wiki.generaloutline.com/share/1c922644-40d8-41fe-98f9-df2b67239d45

SSL_KEY=

SSL_CERT=


# If using a Cloudfront/Cloudflare distribution or similar it can be set below.

# This will cause paths to javascript, stylesheets, and images to be updated to

# the hostname defined in CDN_URL. In your CDN configuration the origin server

# should be set to the same as URL.

CDN_URL=


# Auto-redirect to https in production. The default is true but you may set to

# false if you can be sure that SSL is terminated at an external loadbalancer.

FORCE_HTTPS=true


# Have the installation check for updates by sending anonymized statistics to

# the maintainers

ENABLE_UPDATES=true


# How many processes should be spawned. As a reasonable rule divide your servers

# available memory by 512 for a rough estimate

WEB_CONCURRENCY=1


# Override the maxium size of document imports, could be required if you have

# especially large Word documents with embedded imagery

MAXIMUM_IMPORT_SIZE=5120000


# You can remove this line if your reverse proxy already logs incoming http

# requests and this ends up being duplicative

DEBUG=http


# Comma separated list of domains to be allowed to signin to the wiki. If not

# set, all domains are allowed by default when using Google OAuth to signin

ALLOWED_DOMAINS=
```

**Reply by almereyda** [reply]
Thank you for the data!

Can you try editing and encapsulating your paste with triple tick marks, as described in https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code ? This would make reading and understanding the above much easier.

The same can also be done on all other source code pastes, like the logs above, everywhere where Markdown is supported. It's very neat and eases visual reception of the raw text.

**Reply by dheeban-k** [reply]
@almereyda could u please check now?

**Reply by almereyda** [reply]
Uh, thanks. That's much easier to the eye ;) I guess your `.env` file doesn't mention me in the beginning, but good that worked. Will look at it in the next days.

**Comment by dheeban-k** [comment]
This issue is fixed for me @tommoor

**Reply by o1347mat** [reply]
please tell how you solved this problem?

**Reply by dheeban-k** [reply]
@o1347mat I Referred to CORS topic in GCS and wrote a custom json to fix this.
Ref:- [](https://cloud.google.com/storage/docs/cross-origin)
[https://cloud.google.com/storage/docs/configuring-cors](url)

**Reply by o1347mat** [reply]
damm, i have already configured cors for my bucket but anyway in chrome console i get 
**POST https://storage.googleapis.com/bucket-name 400**

**Reply by dheeban-k** [reply]
Can u share me the ENV vars u set for the bucket related params @o1347mat

**Reply by o1347mat** [reply]
cors policy for my bucket looks like 
```
[{"maxAgeSeconds": 3600, "method": ["*"], "origin": ["*"], "responseHeader": ["Content-Type"]}]
```

env for storage looks like 
```
AWS_S3_UPLOAD_BUCKET_URL=https://storage.googleapis.com/
AWS_S3_FORCE_PATH_STYLE=true
AWS_S3_UPLOAD_BUCKET_NAME=bucket-name
AWS_S3_ACL=private
AWS_S3_UPLOAD_MAX_SIZE=104857600
AWS_REGION=us
AWS_ACCESS_KEY_ID=HMAC access key from gcp 
AWS_SECRET_ACCESS_KEY=HMAC secret from gcp 
```

**Reply by dheeban-k** [reply]
Thank u, ur env file looks good
[
  {
    "origin": ["https://your URL ENDPOINT"],
    "responseHeader": ["Content-Type"],
    "method": ["PUT", "POST"],
    "maxAgeSeconds": 3600
  }
]

This is the CORS.json i mentioned earlier, u can run this command "gsutil cors set CORS.json gs://your gcs bucket name"
It should work now. @o1347mat

**Reply by o1347mat** [reply]
same error in chrome console
```
POST https://storage.googleapis.com/my-bucket 400 files.ts:65  
```
cors looks like
```
[{"maxAgeSeconds": 3600, "method": ["PUT", "POST"], "origin": ["https://my-outline-url"], "responseHeader": ["Content-Type"]}]
```
@dheeban-k  tell me please, you create HMAC secret for Service Account or using your personal(your google acc) key and secret?

**Reply by dheeban-k** [reply]
@o1347mat I used service account for generating HMAC keys
